src/auth/src/credentials/mds_credential.rs

```diff
@@ -28,22 +76,88 @@ const METADATA_FLAVOR: &str = "metadata-flavor";
 const METADATA_ROOT: &str = "http://metadata.google.internal/computeMetadata/v1";
 
 pub(crate) fn new() -> Credential {
-    let token_provider = MDSAccessTokenProvider::builder()
-        .endpoint(METADATA_ROOT)
-        .build();
-    Credential {
-        inner: Arc::new(MDSCredential { token_provider }),
-    }
+    MDSCredentialBuilder::default().build()
 }
 
 #[derive(Debug)]
 struct MDSCredential<T>
 where
     T: TokenProvider,
 {
+    quota_project_id: Option<String>,
+    universe_domain: Option<String>,
     token_provider: T,
 }
 
+/// A builder for creating `MDSCredential` instances.
+///
+/// This struct provides an interface for customizing the creation of
+/// `MDSCredential` instances. It allows you to specify the Metadata Service
+/// endpoint, quota project id, scopes, and universe domain.
+#[derive(Debug, Default)]
+pub struct MDSCredentialBuilder {
+    endpoint: Option<String>,
+    quota_project_id: Option<String>,
+    scopes: Option<Vec<String>>,
+    universe_domain: Option<String>,
+}
+
+#[allow(dead_code)]
+impl MDSCredentialBuilder {
+    /// Sets the endpoint that this credential will use.
+    /// # Default
+    ///
+    /// if `endpoint` is not specified then  `http://metadata.google.internal/computeMetadata/v1` is used.
+    pub fn endpoint<S: Into<String>>(mut self, endpoint: S) -> Self {
+        self.endpoint = Some(endpoint.into());
+        self
+    }
+
+    /// Sets the quota project that this credential will use.
+    pub fn quota_project_id<S: Into<String>>(mut self, quota_project_id: S) -> Self {
+        self.quota_project_id = Some(quota_project_id.into());
+        self
+    }
+
+    /// Sets the universe domain that this credential will use.
+    pub fn universe_domain<S: Into<String>>(mut self, universe_domain: S) -> Self {
+        self.universe_domain = Some(universe_domain.into());
+        self
+    }
+
+    /// Sets the scopes that this credential will use.
+    /// # Default
+    ///
+    /// If scopes are not specified, they are fetched from the Metadata Server.
+    pub fn scopes<S: Into<String>>(mut self, scopes: Vec<S>) -> Self {
+        self.scopes = Some(scopes.into_iter().map(|s| s.into()).collect());
+        self
+    }
+
+    /// Builds and returns a `Credential` instance with the configured settings.
+    ///
+    /// # Returns
+    ///
+    /// A new `Credential` instance that will use Metadata Service.
+    pub fn build(self) -> Credential {
+        let endpoint = self.endpoint.clone().unwrap_or(METADATA_ROOT.to_string());
+
+        let token_provider = MDSAccessTokenProvider::builder()
+            .endpoint(endpoint.clone())
```

Comment:
why to clone endpoint again here? we are not using it after this point, we can just transfer ownership?